### ISSUE TITLE
Compile this code, an error has occurred

### ISSUE THREAD

**Issue by electricface** [rust-lang/rust#12883]
Compile this code, an error has occurred
```rust
//file: abc.rs
use std::io::{BufferedStream,File};
fn main(){
	let file = File::open( &Path::new("message.txt"));
	let mut stream = BufferedStream::new(file);
	stream.write("hello,world".as_bytes() );
	stream.flush();
	return;
	let mut buf = [0, ..100 ];
	match stream.read(buf){
		Ok(nread) => println!("Read {} bytes",nread),
		Err(e) => fail!("error reading :{}",e)
	}
}

```

```
abc.rs:8:6: 8:27 warning: unreachable statement, #[warn(unreachable_code)] on by default
abc.rs:8 	let mut buf = [0, ..100 ];
         	    ^~~~~~~~~~~~~~~~~~~~~
abc.rs:5:2: 5:41 warning: unused result which must be used, #[warn(unused_must_use)] on by default
abc.rs:5 	stream.write("hello,world".as_bytes() );
         	^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
abc.rs:6:2: 6:16 warning: unused result which must be used, #[warn(unused_must_use)] on by default
abc.rs:6 	stream.flush();
         	^~~~~~~~~~~~~~
rustc: /build/buildd/rust-nightly-201403130405~a53242a~saucy/src/llvm/lib/IR/Instructions.cpp:1085: void llvm::StoreInst::AssertOK(): Assertion `getOperand(0)->getType() == cast<PointerType>(getOperand(1)->getType())->getElementType() && "Ptr must be a pointer to Val type!"' failed.
已放弃 (核心已转储)

abort core dumped
```

**Comment by steveklabnik** [comment]
OP's code now gives me

```

```rustc: /home/steve/src/rust/src/llvm/lib/IR/Instructions.cpp:1086: void llvm::StoreInst::AssertOK(): Assertion `getOperand(0)->getType() == cast<PointerType>(getOperand(1)->getType())->getElementType() && "Ptr must be a pointer to Val type!"' failed.

**Comment by tomjakubowski** [comment]
That's the exact same error message, for what it's worth.

**Comment by JustAPerson** [comment]
Updated for current Rust (on playpen as of today 29 March 2015):
**tl;dr**: *this bug appears to be fixed*

[@electricface's example](https://github.com/rust-lang/rust/issues/12883#issue-29432769) *compiles fine*:
```rust
//file: abc.rs
use std::io::{BufStream, Read, Write};
use std::fs::File;
use std::path::Path;

fn main(){
    let file = File::open( &Path::new("message.txt")).unwrap();
    let mut stream = BufStream::new(file);
    stream.write("hello,world".as_bytes() );
    stream.flush();
    return;
    let mut buf = &mut [0; 100];
    match stream.read(buf){
        Ok(nread) => println!("Read {} bytes",nread),
        Err(e) => panic!("error reading :{}",e)
    }
}
```

[@alexcrichton's example](https://github.com/rust-lang/rust/issues/12883#issuecomment-37669052) *compiles fine*:
```rust
fn main(){
    return;
    let _buf = [0; 100];
}
```

[@tomjakubowski's example](https://github.com/rust-lang/rust/issues/12883#issuecomment-44764483) *compiles fine* as is.

As @huonw noted, [@RazvanRotari's example](https://github.com/rust-lang/rust/issues/12883#issuecomment-52385501) *compiles fine*:
```rust
fn main() {
    let a = Cons(20, Box::new(Nil));
    let nr:Option<u32> = a.head();
    match nr {
        Some(a) => println!("{}", a),
        None => println!("")
    }
}

use List::*;
enum List<T> {
    Cons(T, Box<List<T>>),
    Nil
}

impl<T> List<T> {
    fn head(self) -> Option<T> {
        match self {
            Cons(nr, _) => Some(nr),
            Nil => None
        }
    }
}
```

[@P1start's example](https://github.com/rust-lang/rust/issues/12883#issuecomment-53695494) *does not compile* because of borrowck (i.e. unrelated to this bug):
```rust
fn main() {
    let x: &[isize] = &[1, 2, 3];
    let _x: &[[isize]] = &[*x, *x];
}
```
```
<anon>:3:28: 3:30 error: cannot move out of borrowed content
<anon>:3     let _x: &[[isize]] = &[*x, *x];
                                    ^~
<anon>:3:32: 3:34 error: cannot move out of borrowed content
<anon>:3     let _x: &[[isize]] = &[*x, *x];
                                        ^~
<anon>:3:28: 3:30 error: cannot move a value of type [isize]: the size of [isize] cannot be statically determined [E0161]
<anon>:3     let _x: &[[isize]] = &[*x, *x];
                                    ^~
<anon>:3:32: 3:34 error: cannot move a value of type [isize]: the size of [isize] cannot be statically determined [E0161]
<anon>:3     let _x: &[[isize]] = &[*x, *x];
                                        ^~
error: aborting due to 4 previous errors
playpen: application terminated with error code 101
```

[@blastrock's example](https://github.com/rust-lang/rust/issues/12883#issuecomment-55145891) *compiles fine*:
```rust
trait T {
    fn get_t(&mut self) -> Box<T>;
}

struct S;
impl T for S {
    fn get_t(&mut self) -> Box<T> {
        return Box::new(S)
    }
}

fn main() {}
```

**Comment by alexcrichton** [comment]
Thanks @JustAPerson!
